src/main/java/io/hops/hopsworks/expat/migrations/featurestore/featureview/CreateFeatureViewFromTrainingDataset.java

```diff
@@ -0,0 +1,381 @@
+package io.hops.hopsworks.expat.migrations.featurestore.featureview;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import io.hops.hopsworks.common.featurestore.xattr.dto.FeatureViewXAttrDTO;
+import io.hops.hopsworks.common.featurestore.xattr.dto.FeaturegroupXAttr;
+import io.hops.hopsworks.common.featurestore.xattr.dto.FeaturestoreXAttrsConstants;
+import io.hops.hopsworks.common.provenance.core.dto.ProvCoreDTO;
+import io.hops.hopsworks.common.provenance.core.dto.ProvTypeDTO;
+import io.hops.hopsworks.expat.db.dao.hdfs.inode.ExpatHdfsInode;
+import io.hops.hopsworks.expat.migrations.MigrationException;
+import io.hops.hopsworks.expat.migrations.RollbackException;
+import io.hops.hopsworks.expat.migrations.projects.util.XAttrException;
+import io.hops.hopsworks.expat.migrations.projects.util.XAttrHelper;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.eclipse.persistence.jaxb.JAXBContextFactory;
+import org.eclipse.persistence.jaxb.MarshallerProperties;
+import org.eclipse.persistence.oxm.MediaType;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class CreateFeatureViewFromTrainingDataset extends FeatureStoreMigration {
+
+  private final static String GET_ALL_TRAINING_DATASETS =
+      "SELECT a.id, a.feature_store_id, a.name, a.version, a.external_training_dataset_id, b.project_id, " +
+          "d.uid, c.projectname, d.username, d.email, a.created, a.creator, a.description " +
+          "FROM training_dataset AS a " +
+          "INNER JOIN feature_store AS b ON a.feature_store_id = b.id " +
+          "INNER JOIN project AS c ON b.project_id = c.id " +
+          "INNER JOIN users AS d on a.creator = d.uid " +
+          "WHERE a.feature_view_id is null";
+  private final static String GET_ALL_FEATURE_VIEWS =
+      "SELECT fv.id, p.projectname, fv.version " +
+          "FROM feature_view AS fv " +
+          "INNER JOIN feature_store AS fs ON fv.feature_store_id = fs.id " +
+          "INNER JOIN project AS p ON fs.project_id = p.id";
+  private final static String GET_FEATURES =
+      "SELECT fg.name AS fg_name, fg.version AS fg_version, tdf.name AS name " +
+          "FROM training_dataset_feature AS tdf " +
+          "INNER JOIN feature_group AS fg ON tdf.feature_group = fg.id " +
+          "WHERE %s";
```

Comment:
Why are we doing this with string concatenation instead of using just prepared statement parameters?

Reply:
Because there are multiple training_dataset_feature id.